### PR TITLE
[Flink] Support Array type in Flink connector

### LINKED ISSUE #1978: Support Array type in Flink connector
### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss/issues) and found nothing similar.

### Motivation

Enable Flink connector to read and write Array, NestedRow, and Map types for integration with Flink SQL and DataStream API.

### Solution

- Add type mapping between Flink types and Fluss complex types
- Implement serialization/deserialization for complex types in Flink connector
- Support complex types in both source and sink connectors
- Add integration tests for Flink connector with complex types

### Anything else?

_No response_

### Willingness to contribute

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by XuQianJin-Stars**
### Purpose

Linked issue: close #1978

This PR adds comprehensive support for Array type in the Flink connector, enabling users to read and write array data between Flink and Fluss tables. This addresses a key limitation in the connector's type support system.

### Brief change log

- **Core Array type support in common module**:
  - Enhanced Arrow format readers and writers (`ArrowReader`, `ArrowWriter`, `ArrowArrayWriter`, `ArrowFieldWriter`) to handle array serialization and deserialization
  - Updated `ArrowUtils` to support Array type schema conversion
  
- **Flink connector integration**:
  - Implemented bidirectional type conversion in `FlussRowToFlinkRowConverter` and `FlinkRowToFlussRowConverter` to support Array types
  - Added `FlinkAsFlussArray` wrapper class to convert Flink's `ArrayData` to Fluss's `InternalArray`
  - Updated `FlinkAsFlussRow` to support array field access
  
- **Comprehensive test coverage**:
  - Added `FlinkArrayTypeITCase` as the base test class with comprehensive integration tests covering:
    - Arrays of primitive types (int, bigint, string, boolean, double)
    - Arrays with null elements
    - Nested arrays (array of arrays)
    - Arrays in primary key tables
    - Array operations and access
  - Created version-specific test implementations for Flink 1.18, 1.19, 1.20, and 2.1

### Tests

**Unit Tests:**
- Arrow array writer and reader tests for serialization/deserialization

**Integration Tests:**
- `Flink118ComplexTypeITCase`: Array type support for Flink 1.18
- `Flink119ComplexTypeITCase`: Array type support for Flink 1.19
- `Flink120ComplexTypeITCase`: Array type support for Flink 1.20
- `Flink21ComplexTypeITCase`: Array type support for Flink 2.1

Test scenarios include:
- `testArrayOfPrimitiveTypesInLogTable`: Verifies arrays of int, bigint, string, boolean, and double types
- `testArrayWithNullElements`: Ensures proper handling of null elements within arrays and null arrays
- `testNestedArrays`: Tests multi-dimensional arrays (array of arrays)
- `testArrayInPrimaryKeyTable`: Validates array support in tables with primary keys
- `testArrayWithAllDataTypes`: Comprehensive test with all supported data types in arrays
- `testArrayAccessAndCardinality`: Tests array operations and element access

### API and Format

**API Changes:**
- No breaking API changes
- Array type now supported in Flink connector's type mapping

**Storage Format:**
- Enhanced Arrow format to support Array type serialization
- Backward compatible with existing storage formats

### Documentation

**Feature Introduction:**
- This change introduces Array type support as a new feature in the Flink connector
- Users can now create tables with array columns and perform read/write operations through Flink SQL

**Documentation Updates Needed:**
- Update `website/docs/engine-flink/getting-started.md` to reflect Array type support in the type mapping table (change from "Not supported" to "Supported")
- Add examples of using Array types in Flink SQL with Fluss tables

### CODE REVIEW

File: fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/sink/FlinkArrayTypeITCase.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    @Test
+    void testArrayWithAllDataTypes() throws Exception {
+        tEnv.executeSql(
+                "create table array_all_types_test ("
+                        + "id int, "
+                        + "int_array array<int>, "
+                        + "bigint_array array<bigint>, "
+                        + "double_array array<double>, "
+                        + "boolean_array array<boolean>, "
+                        + "string_array array<string>"
```
Comment (leekeiabstraction) [inline fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/sink/FlinkArrayTypeITCase.java]: Are the types here exhaustive? E.g. float, smallint, tinyint, timestamp etc. isn't tested.

https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/types/
https://fluss.apache.org/docs/next/table-design/data-types/

**Comment by leekeiabstraction** [comment]
Integ tests are failing with stack traces that seem related to Arrow*Writer changes

```
Error:  org.apache.fluss.server.kv.KvTabletTest.testPartialUpdateAndDelete  Time elapsed: 0.043 s  <<< ERROR!
java.lang.IndexOutOfBoundsException: index: 0, length: 1 (expected: range(0, 0))
	at org.apache.fluss.shaded.arrow.org.apache.arrow.memory.ArrowBuf.checkIndexD(ArrowBuf.java:319)
	at org.apache.fluss.shaded.arrow.org.apache.arrow.memory.ArrowBuf.chk(ArrowBuf.java:306)
	at org.apache.fluss.shaded.arrow.org.apache.arrow.memory.ArrowBuf.getByte(ArrowBuf.java:508)
	at org.apache.fluss.shaded.arrow.org.apache.arrow.vector.BitVectorHelper.setBit(BitVectorHelper.java:82)
	at org.apache.fluss.shaded.arrow.org.apache.arrow.vector.IntVector.set(IntVector.java:160)
	at org.apache.fluss.row.arrow.writers.ArrowIntWriter.doWrite(ArrowIntWriter.java:38)
	at org.apache.fluss.row.arrow.writers.ArrowFieldWriter.write(ArrowFieldWriter.java:59)
	at org.apache.fluss.row.arrow.ArrowWriter.writeRow(ArrowWriter.java:201)
	```

**Comment by XuQianJin-Stars** [comment]
hi @leekeiabstraction Don't review this PR before the CI passes – I will continue to make revisions.

File: fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/sink/FlinkArrayTypeITCase.java [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+import java.util.Map;
+
+import static org.apache.fluss.flink.FlinkConnectorOptions.BOOTSTRAP_SERVERS;
+import static org.apache.fluss.flink.source.testutils.FlinkRowAssertionsUtils.assertResultsIgnoreOrder;
+import static org.apache.fluss.flink.source.testutils.FlinkRowAssertionsUtils.collectRowsWithTimeout;
+import static org.apache.fluss.server.testutils.FlussClusterExtension.BUILTIN_DATABASE;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Integration tests for Array type support in Flink connector. */
+abstract class FlinkArrayTypeITCase extends AbstractTestBase {
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/sink/FlinkArrayTypeITCase.java]: IT case is very heavy, we should avoid add too many tests just for one purpose.  I suggest to do the following updates:

1. Rename `FlinkArrayTypeITCase` to `FlinkComplexTypeITCase` to cover future Map and Row types. 
2. Add a test for LOG TABLE that covers all array types (ARRAY<not null primitive types>, ARRAY<nullable element types>, several ARRAY nested types). Writing reocords (+ null elements if the element type is nullable) into the table, and read from the table. 
3. Add a test for PK TABLE that is the same with above types. Also writing and reading from the table. But we should test updating and deleting as well. Take `org.apache.fluss.flink.source.FlinkTableSourceITCase#testReadKvTableWithScanStartupModeEqualsFull` as an example about how to verify reading kv table for both snapshot read and incremental read. Then test lookup join the table. 
4. Add exception test that Array type can't be as primary key, or bucket key, or partition key.  Add tests to verify this.

File: fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/sink/FlinkArrayTypeITCase.java [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    }
+
+    @AfterEach
+    void after() {
+        tEnv.useDatabase(BUILTIN_DATABASE);
+        tEnv.executeSql(String.format("drop database %s cascade", DEFAULT_DB));
+    }
+
+    @Test
+    void testSimpleLogTableWithSinkAPI() throws Exception {
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/sink/FlinkArrayTypeITCase.java]: This test is not about array type and is not needed.

File: fluss-common/src/main/java/org/apache/fluss/row/serializer/ArraySerializer.java [hunk 4]
```diff
@@ -113,7 +113,7 @@ public BinaryArray toAlignedArray(InternalArray from) {
             }
             reuseWriter.complete();
 
-            return reuseArray;
+            return reuseArray.copy();
```
Comment (wuchong) [inline fluss-common/src/main/java/org/apache/fluss/row/serializer/ArraySerializer.java]: We should avoid `copy()`, this introduces performance regression.

File: fluss-common/src/main/java/org/apache/fluss/row/encode/CompactedRowEncoder.java [hunk 5]
```diff
@@ -60,8 +60,12 @@ public void encodeField(int pos, Object value) {
 
     @Override
     public CompactedRow finishRow() {
+        int rowSize = writer.position();
+        byte[] rowBytes = new byte[rowSize];
+        writer.segment().get(0, rowBytes, 0, rowSize);
+
         CompactedRow row = new CompactedRow(fieldDataTypes.length, compactedRowDeserializer);
-        row.pointTo(writer.segment(), 0, writer.position());
+        row.pointTo(org.apache.fluss.memory.MemorySegment.wrap(rowBytes), 0, rowSize);
```
Comment (wuchong) [inline fluss-common/src/main/java/org/apache/fluss/row/encode/CompactedRowEncoder.java]: We should avoid deep copy, this introduces performance regression.

File: fluss-common/src/main/java/org/apache/fluss/row/decode/CompactedRowDecoder.java [hunk 6]
```diff
@@ -40,8 +40,12 @@ public CompactedRow decode(byte[] values) {
 
     @Override
     public CompactedRow decode(MemorySegment segment, int offset, int sizeInBytes) {
+        // Copy the data to avoid sharing the underlying MemorySegment
+        byte[] rowBytes = new byte[sizeInBytes];
+        segment.get(offset, rowBytes, 0, sizeInBytes);
+
         CompactedRow compactedRow = new CompactedRow(fieldDataTypes.length, deserializer);
-        compactedRow.pointTo(segment, offset, sizeInBytes);
+        compactedRow.pointTo(MemorySegment.wrap(rowBytes), 0, sizeInBytes);
```
Comment (wuchong) [inline fluss-common/src/main/java/org/apache/fluss/row/decode/CompactedRowDecoder.java]: We should avoid deep copy, this introduces performance regression.

File: fluss-common/src/main/java/org/apache/fluss/record/DefaultLogRecordBatch.java [hunk 7]
```diff
@@ -223,7 +223,10 @@ public CloseableIterator<LogRecord> records(ReadContext context) {
                         rowType,
                         context.getVectorSchemaRoot(schemaId),
                         context.getBufferAllocator(),
-                        timestamp);
+                        timestamp,
+                        context instanceof LogRecordReadContext
+                                ? (LogRecordReadContext) context
+                                : null);
```
Comment (wuchong) [inline fluss-common/src/main/java/org/apache/fluss/record/DefaultLogRecordBatch.java]: This force casting is hack and error-prone in the future. Because this may break if we introduce another ReadContext implementation, and here will use `null`.

File: fluss-common/src/main/java/org/apache/fluss/record/LogRecordReadContext.java [hunk 8]
```diff
@@ -164,6 +168,7 @@ private LogRecordReadContext(
         this.bufferAllocator = bufferAllocator;
         this.selectedFieldGetters = selectedFieldGetters;
         this.projectionPushDowned = projectionPushDowned;
+        this.batchRoots = Collections.synchronizedList(new ArrayList<>());
```
Comment (wuchong) [inline fluss-common/src/main/java/org/apache/fluss/record/LogRecordReadContext.java]: `synchronizedList` is performant bad, please avoid to use this.

File: fluss-common/src/main/java/org/apache/fluss/utils/ArrowUtils.java [hunk 9]
```diff
@@ -159,28 +159,94 @@ public static ArrowReader createArrowReader(
...
         } catch (IOException e) {
             throw new RuntimeException("Failed to deserialize ArrowRecordBatch.", e);
         }
     }
 
+    /**
+     * Fixes writerIndex for all buffers in all vectors after VectorLoader.load().
+     * VectorLoader.load() sets the capacity but not the writerIndex for buffers.
+     */
+    private static void fixVectorBuffers(VectorSchemaRoot schemaRoot) {
```
Comment (wuchong) [inline fluss-common/src/main/java/org/apache/fluss/utils/ArrowUtils.java]: Why do we need this for array type? Why we don't need this before?

File: fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/sink/serializer/RowDataSerializationSchema.java [hunk 10]
```diff
@@ -64,12 +73,25 @@ public RowDataSerializationSchema(boolean isAppendOnly, boolean ignoreDelete) {
     @Override
     public void open(InitializationContext context) throws Exception {
         this.converter = new FlinkAsFlussRow();
+        // For primary key tables (non-append-only), we need to encode the row immediately
+        // to avoid issues with Flink reusing RowData objects
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/sink/serializer/RowDataSerializationSchema.java]: I don't understand this, if it's for primary key tables, then `UpsertWriter` will materialize/encode the row into binary format which already avoid reuse object problems.

**Review by wuchong — CHANGES_REQUESTED** [review]

File: fluss-common/src/main/java/org/apache/fluss/row/InternalRow.java [hunk 11]
```diff
@@ -228,6 +232,57 @@ static FieldGetter createFieldGetter(DataType fieldType, int fieldPos) {
...
+     * position. It returns new objects (GenericArray/GenericMap/GenericMap) for nested
+     * array/map/row types.
+     *
+     * <p>NOTE: Currently, it is only used for deep copying {@link ColumnarRow} for Arrow which
+     * avoid the arrow buffer is released before accessing elements. It doesn't deep copy STRING and
+     * BYTES types, because {@link ColumnarRow} already deep copies the bytes, see {@link
+     * VectorizedColumnBatch#getString(int, int)}. This can be removed once we supports object reuse
+     * for Arrow {@link ColumnarRow}, see {@code CompletedFetch#toScanRecord(LogRecord)}.
+     */
+    static FieldGetter createDeepFieldGetter(DataType fieldType, int fieldPos) {
```
Comment (XuQianJin-Stars) [inline fluss-common/src/main/java/org/apache/fluss/row/InternalRow.java:246]: Good job!
